test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs

```diff
@@ -694,6 +695,136 @@ public async Task ShouldExportActivityWithOneOrMoreExceptionFilters(int mode)
+                    .Build();
+            }
+
+            // Arrange
+            using (var client = this.factory
+                .WithWebHostBuilder(builder =>
+                    builder.ConfigureTestServices(ConfigureTestServices))
+                .CreateClient())
+            {
+                using var request = new HttpRequestMessage(HttpMethod.Get, "/api/values");
+
+                // Act
+                using var response = await client.SendAsync(request);
+            }
+
+            Assert.Equal(1, numberOfCustomCallbacks);
+            Assert.Equal(expectedCustomEventName, actualCustomEventName);
+        }
+
+        [Fact]
+        public async Task DiagnosticSourceExceptionCallbackIsReceivedForUnHandledException()
+        {
+            int numberOfExceptionCallbacks = 0;
+            void ConfigureTestServices(IServiceCollection services)
+            {
+                this.tracerProvider = Sdk.CreateTracerProviderBuilder()
+                    .AddAspNetCoreInstrumentation(new AspNetCoreInstrumentation(
+                        new TestHttpInListener(new AspNetCoreInstrumentationOptions())
+                        {
+                            OnExceptionCallback = (activity, payload) =>
+                            {
+                                numberOfExceptionCallbacks++;
+                            },
+                        }))
+                    .Build();
+            }
+
+            // Arrange
+            using (var client = this.factory
+                .WithWebHostBuilder(builder =>
+                    builder.ConfigureTestServices(ConfigureTestServices))
+                .CreateClient())
+            {
+                try
+                {
+                    using var request = new HttpRequestMessage(HttpMethod.Get, "/api/error");
+
+                    // Act
+                    using var response = await client.SendAsync(request);
+                }
+                catch
+                {
+                    // ignore exception
+                }
+            }
+
+            Assert.Equal(1, numberOfExceptionCallbacks);
+        }
+
+        [Fact(Skip = "Pending Changes https://github.com/open-telemetry/opentelemetry-dotnet/issues/3495")]
+        public async Task DiagnosticSourceExceptionCallBackIsNotReceivedForExceptionsHandledInMiddleware()
+        {
+            int numberOfExceptionCallbacks = 0;
+
+            // configure SDK
+            using var tracerprovider = Sdk.CreateTracerProviderBuilder()
+                .AddAspNetCoreInstrumentation(new AspNetCoreInstrumentation(
+                    new TestHttpInListener(new AspNetCoreInstrumentationOptions())
+                    {
+                        OnExceptionCallback = (activity, payload) =>
+                        {
+                            numberOfExceptionCallbacks++;
+                        },
+                    }))
+                    .Build();
+
+            var builder = WebApplication.CreateBuilder();
```

Comment:
Trying this pattern as it is much simpler to set up.

Reply:
I like it.

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/3394/files